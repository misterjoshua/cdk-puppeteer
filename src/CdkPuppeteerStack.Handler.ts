import { randomUUID } from 'node:crypto';
import { GetObjectCommand, S3 } from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import chromium from '@sparticuz/chromium';

import type * as lambda from 'aws-lambda';
import puppeteer from 'puppeteer';

// Configure the output location.
const outputBucketName = process.env.OUTPUT_BUCKET_NAME!;
const outputBucketClient = new S3({
  region: process.env.OUTPUT_BUCKET_REGION!,
});

export const handler: lambda.APIGatewayProxyHandlerV2 = async () => {
  // Launch puppeteer with sparticuz's chromium distro.
  const browser = await puppeteer.launch({
    args: chromium.args,
    defaultViewport: chromium.defaultViewport,
    executablePath: await chromium.executablePath(),
    headless: chromium.headless,
  });

  try {
    // Get a screenshot
    const page = await browser.newPage();
    await page.goto('https://aws.amazon.com/');
    const screenshot = await page.screenshot();

    // Write it to s3
    const outputKey = `temp-output/${randomUUID()}.png`;
    await outputBucketClient.putObject({
      Bucket: outputBucketName,
      Key: outputKey,
      Body: screenshot,
      ContentType: 'image/png',
    });

    // Generate a pre-signed URL to download the screenshot
    const downloadUrl = await getSignedUrl(outputBucketClient, new GetObjectCommand({
      Bucket: outputBucketName,
      Key: outputKey,
    }), { expiresIn: 60 * 60 });

    // Return the URL
    return {
      statusCode: 200,
      body: JSON.stringify({
        downloadUrl,
      }),
    };
  } finally {
    await browser.close();
  }
};