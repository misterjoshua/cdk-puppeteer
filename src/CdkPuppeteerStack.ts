import { CfnOutput, Duration, RemovalPolicy, Stack, StackProps } from 'aws-cdk-lib';
import { FunctionUrl, FunctionUrlAuthType, HttpMethod, InvokeMode, Runtime } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

export class CdkPuppeteerStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    // Create an S3 bucket to store the screenshots
    const outputBucket = new Bucket(this, 'OutputBucket', {
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    // Automatically delete objects after 30 days
    outputBucket.addLifecycleRule({
      expiration: Duration.days(30),
      prefix: 'temp-output',
    });

    // Create a Lambda function that takes a screenshot of a website
    const handler = new NodejsFunction(this, 'Handler', {
      runtime: Runtime.NODEJS_LATEST,
      memorySize: 8192,
      timeout: Duration.seconds(30),
      bundling: {
        // This module is required but doesn't survive bundling.
        nodeModules: ['@sparticuz/chromium'],
      },
    });

    // Let the handler read/write to the output bucket
    outputBucket.grantReadWrite(handler);
    handler.addEnvironment('OUTPUT_BUCKET_NAME', outputBucket.bucketName);
    handler.addEnvironment('OUTPUT_BUCKET_REGION', this.region);

    // Create a URL to invoke the Lambda function
    const functionUrl = new FunctionUrl(this, 'FunctionUrl', {
      function: handler,
      invokeMode: InvokeMode.BUFFERED,
      authType: FunctionUrlAuthType.NONE,
      cors: {
        allowedOrigins: ['*'],
        allowedHeaders: ['*'],
        allowedMethods: [HttpMethod.ALL],
      },
    });

    // Display the URL in the CloudFormation outputs.
    new CfnOutput(this, 'Url', {
      value: functionUrl.url,
    });
  }
}
