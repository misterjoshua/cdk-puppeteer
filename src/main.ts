import { App } from 'aws-cdk-lib';
import { CdkPuppeteerStack } from './CdkPuppeteerStack';

const app = new App();

new CdkPuppeteerStack(app, 'cdk-puppeteer-dev', {
  // for development, use account/region from cdk cli
  env: {
    account: process.env.CDK_DEFAULT_ACCOUNT,
    region: process.env.CDK_DEFAULT_REGION,
  },
});

app.synth();