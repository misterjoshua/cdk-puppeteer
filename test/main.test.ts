import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

import { CdkPuppeteerStack } from '../src/CdkPuppeteerStack';

test('Snapshot', () => {
  const app = new App();
  const stack = new CdkPuppeteerStack(app, 'test');

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});