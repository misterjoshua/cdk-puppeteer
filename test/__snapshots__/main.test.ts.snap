// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Outputs": {
    "Url": {
      "Value": {
        "Fn::GetAtt": [
          "FunctionUrl84B92DCA",
          "FunctionUrl",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": {
      "DependsOn": [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "6c1e9b465fa4b2d651dbc9ce3e732d8702a7b19137327684a71d89f1d355f1a2.zip",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Lambda function for auto-deleting objects in ",
              {
                "Ref": "OutputBucket7114EB27",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "FunctionUrl84B92DCA": {
      "Properties": {
        "AuthType": "NONE",
        "Cors": {
          "AllowHeaders": [
            "*",
          ],
          "AllowMethods": [
            "*",
          ],
          "AllowOrigins": [
            "*",
          ],
        },
        "InvokeMode": "BUFFERED",
        "TargetFunctionArn": {
          "Fn::GetAtt": [
            "Handler886CB40B",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Url",
    },
    "Handler886CB40B": {
      "DependsOn": [
        "HandlerServiceRoleDefaultPolicyCBD0CC91",
        "HandlerServiceRoleFCDC14AE",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "baeb8ef463a520dd1e5cc49b6b67d778dbe3d1432748df3053cb1778b096195f.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "OUTPUT_BUCKET_NAME": {
              "Ref": "OutputBucket7114EB27",
            },
            "OUTPUT_BUCKET_REGION": {
              "Ref": "AWS::Region",
            },
          },
        },
        "Handler": "index.handler",
        "MemorySize": 8192,
        "Role": {
          "Fn::GetAtt": [
            "HandlerServiceRoleFCDC14AE",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "HandlerServiceRoleDefaultPolicyCBD0CC91": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "OutputBucket7114EB27",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "OutputBucket7114EB27",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "HandlerServiceRoleDefaultPolicyCBD0CC91",
        "Roles": [
          {
            "Ref": "HandlerServiceRoleFCDC14AE",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "HandlerServiceRoleFCDC14AE": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "HandlerinvokefunctionurlEFA6983A": {
      "Properties": {
        "Action": "lambda:InvokeFunctionUrl",
        "FunctionName": {
          "Fn::GetAtt": [
            "Handler886CB40B",
            "Arn",
          ],
        },
        "FunctionUrlAuthType": "NONE",
        "Principal": "*",
      },
      "Type": "AWS::Lambda::Permission",
    },
    "OutputBucket7114EB27": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "LifecycleConfiguration": {
          "Rules": [
            {
              "ExpirationInDays": 30,
              "Prefix": "temp-output",
              "Status": "Enabled",
            },
          ],
        },
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "OutputBucketAutoDeleteObjectsCustomResource6C7A161F": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "OutputBucketPolicyAE009DDC",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "OutputBucket7114EB27",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "OutputBucketPolicyAE009DDC": {
      "Properties": {
        "Bucket": {
          "Ref": "OutputBucket7114EB27",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:PutBucketPolicy",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "OutputBucket7114EB27",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "OutputBucket7114EB27",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
